src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java

```diff
@@ -0,0 +1,201 @@
+package fi.aalto.cs.apluscourses.intellij.actions;
+
+import com.intellij.openapi.actionSystem.AnAction;
+import com.intellij.openapi.actionSystem.AnActionEvent;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.application.ex.ApplicationEx;
+import com.intellij.openapi.project.DumbAware;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.ui.Messages;
+import fi.aalto.cs.apluscourses.intellij.model.APlusProject;
+import fi.aalto.cs.apluscourses.intellij.model.IntelliJModelFactory;
+import fi.aalto.cs.apluscourses.intellij.services.MainViewModelProvider;
+import fi.aalto.cs.apluscourses.intellij.services.PluginSettings;
+import fi.aalto.cs.apluscourses.intellij.utils.SettingsUtil;
+import fi.aalto.cs.apluscourses.model.Course;
+import fi.aalto.cs.apluscourses.model.MalformedCourseConfigurationFileException;
+import fi.aalto.cs.apluscourses.model.UnexpectedResponseException;
+import fi.aalto.cs.apluscourses.presentation.CourseViewModel;
+import fi.aalto.cs.apluscourses.ui.IntelliJDialogs;
+import fi.aalto.cs.apluscourses.ui.base.Dialogs;
+import fi.aalto.cs.apluscourses.utils.CoursesClient;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import org.apache.commons.io.FileUtils;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class CourseProjectAction extends AnAction implements DumbAware {
+
+  @NotNull
+  private MainViewModelProvider mainViewModelProvider;
+
+  @NotNull
+  private Dialogs dialogs;
+
+  public CourseProjectAction() {
+    this(PluginSettings.getInstance(), new IntelliJDialogs());
+  }
+
+  /**
+   * Construct a course project action with the given main view model provider and dialogs.
+   */
+  public CourseProjectAction(@NotNull MainViewModelProvider mainViewModelProvider,
+                             @NotNull Dialogs dialogs) {
+    this.mainViewModelProvider = mainViewModelProvider;
+    this.dialogs = dialogs;
+  }
+
+  @Override
+  public void actionPerformed(@NotNull AnActionEvent e) {
```

Comment:
Nit: it might be more convenient e.g. for testing purposes to break this longer method into a smaller chunks